tech/firewall-rule-format.md

```diff
@@ -575,15 +600,487 @@ For the record, the reason why the protocol is `*` in Azure is that
+      "name": "Port_22",
+      "etag": "W/\"7a502bb3-4f77-4c23-a5fa-1fda055b021d\""
+    }
+
+To take multiple CIDRs into account and finding whether their union is
+`0.0.0.0/0` should perhaps be outside the scope of MVP. Detecting this
+kind of corner cases (which could be exploited as clever workarounds to
+avoid detection) should perhaps be picked up as enhancements later.
+
+Until then, perhaps to be extra safe we should look for `0.0.0.0/0` as
+one of the entries in the `source_address_prefixes` list.
+
+
+#### `source_address_prefix` set to `Internet`
+
+Yet another way to expose a port to the Internet is to use a service
+tag. This security rule was created by choosing *Service Tag* from the
+*Source* dropdown menu and choosing *Internet* from the *Source service
+tag* dropdown menu.
+
+    {
+      "id": "/subscriptions/a8daa361-b951-46b5-8e9d-b375f2ffe9fd/resourceGroups/rg1/providers/Microsoft.Network/networkSecurityGroups/vm1-nsg/securityRules/Port_22",
+      "protocol": "*",
+      "source_port_range": "*",
+      "destination_port_range": "22",
+      "source_address_prefix": "Internet",
+      "source_address_prefixes": [],
+      "destination_address_prefix": "*",
+      "destination_address_prefixes": [],
+      "source_port_ranges": [],
+      "destination_port_ranges": [],
+      "access": "Allow",
+      "priority": 100,
+      "direction": "Inbound",
+      "provisioning_state": "Succeeded",
+      "name": "Port_22",
+      "etag": "W/\"a9ceecf4-eb4a-4fb8-b961-4cd9ab246e70\""
+    }
+
+
+### Any Source: GCP
+
+GCP contains this firewall rule to allow connectivity port 22 from
+anywhere as one of the default rules:
+
+    {
+      "kind": "compute#firewall",
+      "id": "6979850877904078814",
+      "creationTimestamp": "2018-12-04T04:55:13.187-08:00",
+      "name": "default-allow-ssh",
+      "description": "Allow SSH from anywhere",
+      "network": "https://www.googleapis.com/compute/v1/projects/strong-augury-224506/global/networks/default",
+      "priority": 65534,
+      "sourceRanges": [
+        "0.0.0.0/0"
+      ],
+      "allowed": [
+        {
+          "IPProtocol": "tcp",
+          "ports": [
+            "22"
+          ]
+        }
+      ],
+      "direction": "INGRESS",
+      "logConfig": {
+        "enable": false
+      },
+      "disabled": false,
+      "selfLink": "https://www.googleapis.com/compute/v1/projects/strong-augury-224506/global/firewalls/default-allow-ssh"
+    }
+
+Here is a how a custom firewall rule to allow connectivity to TCP port
+8000 with *Source IP ranges* as `0.0.0.0/0` looks like:
+
+    {
+      "kind": "compute#firewall",
+      "id": "4387657442666205482",
+      "creationTimestamp": "2019-03-26T11:30:29.863-07:00",
+      "name": "allow-8000",
+      "description": "",
+      "network": "https://www.googleapis.com/compute/v1/projects/strong-augury-224506/global/networks/default",
+      "priority": 1000,
+      "sourceRanges": [
+        "0.0.0.0/0"
+      ],
```

Comment:
I was just wondering, is the case for multiple CIDRs not applicable for GCP?

Reply:
It is still applicable. The reason why we have `"source_address_prefixes": ["0.0.0.0/1", "128.0.0.0/1"]` in the Azure example but we only have `"sourceRanges": ["0.0.0.0/0"]` in the GCP example is that the Azure portal did not let me enter `0.0.0.0/0` as a valid CIDR at all, so just for the sake of writing an example I entered two CIDRs that when combined is equivalent to `0.0.0.0/0`.

Having said that, I am not planning to perform union of CIDRs to see if it is equivalent to the entire Internet in the initial implementation. That is a very interesting problem to solve though.